ios in tinygrad

roryclear opened the pull request.

roryclear committed:
starting again, inline and remove stuff

roryclear committed:
start new msg_ios

roryclear committed:
object or class for ptr

roryclear committed:
add more to ios

roryclear committed:
add more ios to metal without breaking either

roryclear committed:
add more ios without breaking metal

roryclear committed:
more ios, metal still works

roryclear committed:
inline an as_buffer call

roryclear committed:
add delete and fix copyin

roryclear committed:
fix offset

roryclear committed:
start jit, again

roryclear committed:
jit might work now

roryclear committed:
add if for jit, remove size on copyout

roryclear committed:
add buf view from ios

roryclear committed:
start removing metal crutch

roryclear committed:
move functions to device class

roryclear committed:
else if

roryclear committed:
clean ios a bit

roryclear committed:
start batching

roryclear committed:
clean more

roryclear committed:
drop buf_ios name

roryclear committed:
1d buffers in flight

roryclear committed:
drop _ios names

roryclear committed:
fix all_res

roryclear committed:
move to one file

roryclear committed:
add ops ios

roryclear committed:
revert metal

roryclear committed:
revert metal?

roryclear committed:
clean a bit

roryclear committed:
use 1000

roryclear committed:
move copyout idx

roryclear committed:
check size

roryclear committed:
add elapsed_time and ios renderer for search to work

roryclear committed:
remove ref to gpt2 weights

roryclear committed:
use metalrenderer again, only check size on wait for now to get back speed

roryclear committed:
clean

roryclear committed:
delete file after copy to metal

roryclear committed:
add check and error for files

roryclear committed:
send weights to ios if file not in bundle

roryclear committed:
clean file transfer

roryclear committed:
free up memory after copy

roryclear committed:
clean a bit

roryclear committed:
Merge branch 'tinygrad:master' into ios_3

roryclear committed:
Merge branch 'tinygrad:master' into ios_3

roryclear commented on tinygrad-ios/tinygrad-ios/ViewController.m:
all the iOS code is here, everything else is blank

AlexCheema commented on tinygrad/device.py:
```diff
@@ -180,7 +180,8 @@ class CompileError(Exception): pass
 class Compiler:
   def __init__(self, cachekey:Optional[str]=None): self.cachekey = None if getenv("DISABLE_COMPILER_CACHE") else cachekey
-  def compile(self, src:str) -> bytes: return src.encode()   # NOTE: empty compiler is the default
+  def compile(self, src:str) -> bytes: 
+    return src.encode()   # NOTE: empty compiler is the default
```
Formatting change

roryclear replied:
reverted

AlexCheema commented on tinygrad/engine/realize.py:
```diff
@@ -117,6 +117,9 @@ def __init__(self, total_sz, dest_device, src_device):
     else: name = f"{type(self).__name__[6:].lower()} {total_sz:8d}, {dest_device[:7]:>7s} <- {src_device[:7]:7s}"
     super().__init__(colored(name, "yellow"), dest_device, 0, total_sz)
   def copy(self, dest, src):
+    if src.device.startswith("DISK") and dest.device == "IOS": #TODO, hack for iOS
+      dest.allocator.copy_from_disk(dest,src)
+      return
```
Hack?

roryclear replied:
this is needed to be able to copy weights from iOS's bundle to metalbuffers, it made testing this faster, but won't be needed for exo. Can just send weights over the network (which this does if the file isn't found) and then cache them

roryclear replied:
I've removed this, it uses the same path as metal now

AlexCheema commented on tinygrad/runtime/ops_ios.py:
```diff
@@ -0,0 +1,239 @@
+class IOSDevice(Compiled):
+  def __init__(self, device:str):
+    self.device = "d"
+    self.ip = "http://192.168.1.6:8081" #iphone ip address
```
Hardcoded?

roryclear replied:
have changed this to a command line arg

AlexCheema commented on tinygrad/runtime/ops_ios.py:
```diff
@@ -0,0 +1,239 @@
+    self.send_queue()
+    url = self.ip + "/bytes/" + str(size) + "/" + dest
+    status = 400
+    while status != 200:
+      try:
+        headers = {}
+        response = requests.post(url, data=data,headers=headers,timeout=3600)
+        if response.status_code == 200:
+            status = 200
+            if len(response.text) > 0:
+              return response.text
+        else:
+            time.sleep(0.1)
+      except requests.exceptions.RequestException as e:
+        time.sleep(0.2)
```
This error handling seems not great. Sleeping arbitrarily like this will block the main thread potentially indefinitely.
Can we just propagate any errors up?
Might make sense to handle the errors higher up.

roryclear replied:
I've cleaned this code a bit and added a retry limit that throws an error if reached. I don't think sleeping on the main thread here is an issue, the queue is only sent when tinygrad needs data from the iOS device, so nothing can happen until that data is received anyway https://github.com/roryclear/tinygrad/blob/7f9e861117de2f468918e6d9c10afd52c2563b0c/tinygrad/runtime/ops_ios.py#L197

AlexCheema commented on tinygrad/runtime/ops_ios.py:
```diff
@@ -0,0 +1,239 @@
+            time.sleep(0.1)
+      except requests.exceptions.RequestException as e:
+        time.sleep(0.2)
+    return
```
Redundant return?

roryclear replied:
removed

AlexCheema commented on tinygrad/runtime/ops_ios.py:
```diff
@@ -0,0 +1,239 @@
+            time.sleep(0.1)
+      except requests.exceptions.RequestException as e:
+        time.sleep(0.2)
+    return
```
Redundant return?

AlexCheema commented on tinygrad/runtime/ops_ios.py:
```diff
@@ -0,0 +1,239 @@
+    for x in args: req.append(x)
+    if res != None: req.append(res)
+    self.queue.append(req)
+    if selector in ["copyout","maxTotalThreadsPerThreadgroup","elapsed_time","file_exists"] or len(self.queue) > 1000: #todo, don't send bytes as a string etc
```
todo?

roryclear replied:
I had a size limit of 1000 for the queues because the iOS app couldn't handle large requests, I've fixed that now

AlexCheema reviewed (changes requested):
Left some comments.
What's the intended deployment path here?
I think it would make sense to move this code into exo itself since it's unlikely to be merged into tinygrad.

roryclear committed:
remove return

roryclear committed:
restore format

roryclear commented:
> Left some comments. What's the intended deployment path here? I think it would make sense to move this code into exo itself since it's unlikely to be merged into tinygrad.

I'm thinking exo imports its own fork of tinygrad that has extra code for iOS, with at least a couple of changes from this:

- gRPC, I assume this is a hard requirement for exo at the moment? I haven't put much effort into the networking stuff on this because I assume it will be ripped out anyway, just wanted to get something up quickly
- Caching, this repeats a lot of work for every run (like tinygrad does), but can be a lot quicker in a cluster if it knows the same code as before is being ran

https://github.com/roryclear/tinygrad/blob/ios_3/tinygrad-ios/tinygrad-ios/ViewController.m is where the iOS code is if you haven't seen it. It just takes in and runs the instructions sent from tinygrad, then sends back the result, just like a GPU. I did get tinygrad compiling swift and objc code on a previous branch, but that way is much less practical as you have to rebuild your iOS app for every configuration and on every device

roryclear committed:
clean networking

roryclear committed:
clean a bit

roryclear committed:
clean

roryclear committed:
handle larger queues in iOS, remove 1000 size limit

roryclear committed:
remove copy from bundle hack

roryclear committed:
use arg for ip

roryclear committed:
remove outdated notes

roryclear committed:
move ios ip to helpers, it broke other args

roryclear committed:
clean a bit

roryclear committed:
Merge branch 'master' into ios_3

roryclear committed:
update for new convention

roryclear committed:
Merge pull request #2 from tinygrad/master

sync

roryclear committed:
update clprg name

roryclear committed:
sync fork (#3)

* script to run regressed sd conv on metal (#7995)

* script to run regressed sd conv on metal

this and other similar `conv2d + add` kernels contributed to most of the speed regression

* # ruff: noqa: E501

* alu(c?t0:f0, c?t1:f1) -> c?alu(t0,t1):alu(f0,f1) (#7900)

* alu(c?t0:f0, c?t1:f1) -> c?alu(t0,t1):alu(f0,f1)

only do if at least one branch is const, so total alu won't increase

* tests and interesting TODO cases

* add process replay to webgpu tests (#7998)

* hotfix: print only kernel time

* add typing to basicblocks (#7999)

* minor uop speedups [pr] (#8002)

* minor uop cleaner [pr]

* free uop creation speed by removing WeakValueDictionary

* a lil faster

* disable that test

* lines

* and it doesn't print non hit patterns

* test: don't ref until schedule (#8004)

* const_like can return a valid [pr] (#8005)

* const_like can return a valid [pr]

* fixup

* don't mutate the uop/lazybuffer, just the Buffer [pr] (#8000)

* don't mutate the uop/lazybuffer, just the Buffer [pr]

* fix red test

* try different fix

* that

* that's the right fix

* test for fixed behavior

* bump to 3.12

* assert const folding return shape matches [pr] (#8006)

* sacrifice ClangGraph on the altar of lines [pr] (#8009)

* Improved div folding (#7996)

* First version of div_mod folding together

* Working version with old div folding behaviour

* Test is fixed

* Fix linting

* Happy mypy

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* cache indexed uops in st [pr] (#8008)

* cache indexed uops in st [pr]

* remove arg from range

* YoloV8 on WebGPU (#8007)

Port YoloV8 to WebGPU

* add test case of realize_size with 0 in shape (#8011)

* Remove WebGL (#8012)

* clean up wgsl_matcher [pr] (#8015)

use more UPat syntatic sugar and remove unneeded rules

* minor tweak in ptx asm_for_op [pr] (#8017)

always compare with dtypes instead of name string

* minor nn cleanups (#8018)

use more .numel and .ndim

* reduce on wmma (#8016)

* lazy srcs shape mistmatch assert + fix ASSIGN [pr] (#8014)

* lazy srcs shape mistmatch assert [pr]

* duplicate assert

* base it later

* keep the assert

* hcq c-like args state (#8020)

* hcq c-like args state

* ugh

* Dfix

* rename

* i

* unused mul add lt rule [pr] (#8022)

* update uop and tests to not use lt/gt/le/ge [pr] (#8023)

just use dunder methods, eventually remove those from ops

* replace .lt in rewrite rules with < [pr] (#8024)

* update function to not use gt/lt [pr] (#8025)

pr does not test this, but it's the same

* remove lt/gt/le/ge from SimpleMathTrait [pr] (#8027)

just use the dunder methods

* clean up rewrite logic + merge siblings (#8026)

* clean up rewrite logic [pr]

* simpler

* merge sibling blocks

* no PR

* hotfix: early INDEX

* local reordering in block (#8029)

* local reordering in block

* load (and parents) is highest priority

* minor loads in order

* comments

* explicit depth

* simpler

* matters less, but store early too

* Use MTLCompiler directly (v2) (#7920)

* Use MTLCompiler directly (v2)

* to_block_literal and REQUEST_TYPE_COMPILE

* Rewrite command encoding

* Revert to_block_literal

* Maybe that's more readable to some people?

* Typo and comment about stdlib caching

* Update ops_metal.py

* Update ops_metal.py

* Update ops_metal.py

---------

Co-authored-by: George Hotz <72895+geohot@users.noreply.github.com>

* bunch up ops for lines [pr] (#8030)

* tar_extract with Tensors (#7853)

* initial

* USTAR, PAX and GNU support + testing

* from_bytes byteorder

* use TarInfo.frombuf

* tensor only usage

* remove contextlib.suppress

* shorter ow,pax

* more tests

* testing length + move tests

* cleanup

* new approach: RawTensorIO

* fix fetch

* enable read test

* cleanup and ignore fix

* fix for python < 3.12

* make it RawIO

* functions

---------

Co-authored-by: George Hotz <72895+geohot@users.noreply.github.com>
Co-authored-by: chenyu <chenyu@fastmail.com>

* simpler block reorder function [pr] (#8031)

* simpler block reorder function [pr]

* simpler

* block_reorder in substitute, so wasteful otherwise

* extend and count

* leave push logic for same order

* sort new ctx

* less loop

* Revert "less loop"

This reverts commit 30249d097a8affba07a15c0ba1d6b56c9eea3006.

* hotfix: use kernel timings instead of python timings in speed_v_theoretical

* hotfix: don't spam BEAM debug in speed_v_theoretical

* hotfix: bump expectations in speed_v_theoretical

* use BufferedReader (#8032)

* Run WebGPU tests on ubuntu (#8033)

* make device on uop optional [pr] (#8034)

* Fast YoloV8 on WebGPU (#8036)

* Fast yolov8 with downscaled input

* Faster + FPS meter

* Add loader while model is downloading/compiling

* Title touchup

* simple contiguous_while_contiguous prereqs [pr] (#8038)

* simple contiguous_while_contiguous prereqs [pr]

* early realize

* fine if it's folding a non-contig buffer

* vectorized input in div_and_mod_folding returns None [pr] (#8041)

* lower v_theoretical gemv on red (#8042)

tiny7 is still slower https://github.com/tinygrad/tinygrad/actions/runs/12166149038/job/33931736130#step:8:209

* Fold nested div with const (#8010)

* Rebase nested div and with const

* Update the ordering

* return None on vectors

Fixes cpu test

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* simple onnx_ops cleanups (#8003)

* simple clean ups first

* more work

* kinda have adam

* ooo momentum worked nicely

* almost there

* wow.. is the onnx test wrong

* nicer optim stuff

* just skip that test

* small comment changes

* use naming convention from other parts of codebase

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* minor get_grouped_dims cleanup [pr] (#8044)

* use argfix in smax/smin and remove if [pr] (#8045)

* minor push_swizzle_down_through_elementwise cleanup [pr] (#8046)

walrus, and if x are the same, prod(x) must be the same

* hook_overflow -> safe_exp2 [pr] (#8047)

that's the only use case, so no need for indirection

* test cases of always True/False lt (#8048)

* test cases of always True/False lt

* one more

* QwQ-32B-Preview support (#7962)

* load weights with some debugging

* start running a prompt

* cleanup

* optionally permute layers and cleanup

* add validation for simple prompt

* small cleanup

* minor cleanup with formatting download links

* add a longer prompt

* add timing option

* some typings

* remove unused arg

* reset GlobalCounters

* minor cleanups

* accept filename decorator [pr] (#8049)

* accept filename decorator [pr]

* add test for safe_load

* bring old tar tests back

* bump download cache version

* do gpuocelot copy manually [pr] (#8050)

* lower test_gemv_4096_16384 expectations

* reduce folding simple tests [pr] (#8040)

* reduce folding simple tests [pr]

* test for view and realized src pattern

* realize / buffer behavior

* Proper dtypes for input/output of exported WebGPU model (#8053)

* Respect input/output dtypes in exported WebGPU model

* Add some comments about skipped dtypes

* nv one _gpu_alloc (#8055)

* amd general _gpu_alloc (#8056)

* amd general _gpu_alloc

* hmm

* ops

* YoloV8 WebGPU fixes (#8057)

* Bump up input size to 416, show if webgpu is not supported

* Minor fix in export_model

* lower test_conv_3x3_256_32_32_256_256 expectation (#8060)

failed https://github.com/tinygrad/tinygrad/actions/runs/12182799887/job/33982676812#step:9:210

* Simple onnx.py clean ups (#8054)

* start

* simplify ops

* why did this not work before

* will split buffer parse to separate pr

* flip the error order

* only this much for now

* to_python_const clean up

* minimize diff

* move tensor_methods into onnx.py

* improve some type signatures

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* simpler dtypes.max for int [pr] (#8058)

* combine parentless reduce rule [pr] (#8059)

* add _one_hot_along_dim helper for Tensor.arange masking (#8039)

* feelsbadman

* feelsextrabadman

* make sure indices is on same device as self Tensor

* renamed to _one_hot_along_dim

* revert onnx change will do them in onnx only PRs

* address feedback

* add onnx changes here too

* make pad arg better

* revert pad arg

* maybe still keep dim

* simplify onehot onnx ops more

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* Add pattern for div mod recombine with gcd (#8061)

Co-authored-by: chenyu <chenyu@fastmail.com>

* fix typing of UOp.range [pr] (#8062)

start/end should not be float or bool

* minor to_indexed_uops cleanup [pr] (#8063)

* reshape mask cleanups [pr] (#8064)

don't need canonicalize_st because we always merge 1 in `_merge_dims`

* unify expand conditions [pr] (#8065)

same condition (check if old == new or old == 1) in tensor and view. also renamed _pad_left to _align_left because it's not really a pad

* improve idiv _min_max (#8066)

for the cases that the we don't know the exact bounds, we might still know the sign. with this, can remove some resolve for symbolic shapetracker

* refactor safe_load (#8035)

* refactor safe_load

* cleanup

* mypy==1.13.0 (#7990)

* explicit instantiation and narrowing asserts

* explicit cast

* bump

* one line assert

* handle case for no copy_queue_t

* Revert "handle case for no copy_queue_t"

This reverts commit 38347806ca9d1c6433966edc9249479edaea0a33.

* more readable control flow

---------

Co-authored-by: George Hotz <72895+geohot@users.noreply.github.com>

* example: self_tokenize. someday tinygrad will be recursively self improving

* update self_tokenize, fix pylint maybe

* hotfix: actually fix pylint, it's a python 3.10 issue

* test copy inside jit [pr] (#8072)

* fix bug in jit prune with copy [pr] (#8073)

* u32 to f16 in tinygrad (#8074)

* f16 decompression in tinygrad

* Typing and cleanup

* qcom has not transfer (#8075)

* qcom alloc is not hcq alloc

* maybe base?

* test

* combine get inputs and type_parse function in onnx (#8069)

* 1 is simpler than 2

* variable name

* change error wording

* shapes for sequence type must be homogeneous

* clean up reshape size check [pr] (#8067)

removed a resolve, and remove special case for 0 size assert since it's covered by generic size check

* Cleaner rule for mul/idiv by power of two [pr] (#8076)

* Cleaner rule for mul/idiv by power of two

* Change comment

* Revert "combine get inputs and type_parse function in onnx (#8069)" (#8079)

This reverts commit 074a67a6eb25a9dffb10a637707169175ac571e2.

* add ceil_mode for avg_pool and max_pool (#7579)

* wip pool

* check CI for remove alternative implementation

* Revert "check CI for remove alternative implementation"

This reverts commit 7b1bb900e58d592e21e1f25a98316109d9980188.

* fix test

* tests tests tests

* slap a resolve on it

* fix comment

* a little simpler pool

* check CI for removal again

* Revert "check CI for removal again"

This reverts commit be798b7857ea2433d8792e1034884aadae07d4ce.

* small

* update

* some ez tests

* english

* clean up code

* fix ruff

* how did I +25 lines?

* small clean ups

* moar clean ups

* try test_avgpool2d_failure2 in CI

* final clean up

* exclude bug fix

* avg underscore pool

* no more edge case stuff

* add better comments for explanation

* add test cases for decreasing end padding

* address feedback

* improve test coverage

* tiny more polish as we wait for lines :D

* more readable code ordering

* add to documentation

* oops

* set to False instead

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* Use matching JS TypedArray for buffer dtype (#8080)

* free up lines for const_arg [pr] (#8083)

* Use atomicLoad builtin when loading atomic type (#8084)

* move view_supported_devices to device [pr] (#8085)

* skip test_schedule_mem_used_with_inputs [pr] (#8086)

* Move efficientnet example (#8087)

* Fix uint8 cast underflow (#6305)

* hacky fix for cast

* only float to uint8

* limit to float -> uint8

* touchup alu cast test

* improve tests and support more float to unsigned casts

* del one repeated test

* del 1 more repeated test

* try removing expected failure test

* hmmm try 1 more

* skip tests for flakiness

* uint64 super flaky

* clean up

* grammar

* just match numpy

* why is CI numpy different from local numpy

* increase verbosity

* try

* try2

* try3

* try4

* yeah idk

* new direction

* try again

* just don't support uint32 and uint64

* done?

* oops

* comment

* documentation

* it is what it is

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* unrelated test fixups from delete_lazy [pr] (#8088)

* unrelated test fixups from delete_lazy [pr]

* fine if it's scheduled later

* hcq bump alloc (#8078)

* hcq bump alloc

* hm

* nv

* typo

* combine get inputs and type_parse function in onnx [fixed] (#8081)

* 1 is simpler than 2

* variable name

* change error wording

* shapes for sequence type must be homogeneous

* bug fix for model benchmark

* fix comments too

---------

Co-authored-by: chenyu <chenyu@fastmail.com>

* big graph init conceptual cleanup [pr] (#8090)

* keep Ops.BUFFER naming consistent [pr]

* big graph init conceptual cleanup [pr]

* make everything pass through

* pylint doesn't complain now

* fix return type of Tensor.pow (#8091)

int to power of int should return int etc, it hints that we would like to have Ops.POW

* tiny metal cleanup (#8089)

* tiny metal cleanup

* cast

* sry

* remove ordered parents, seems like dead code [pr] (#8092)

* remove ordered parents, seems like dead code

* no need to dedup

* maskless const can lower without valid, p1 [pr] (#8094)

* onnx Bitwise ops (#8095)

free stuff!

* don't FUZZ_ALL_ACTIONS in fuzz_linearizer.py (#8096)

mostly for speed, this is just making sure the script runs

* default tensors to int32 in test_ops (#8097)

torch defaults to int64 but we care more about int32 anyway. remove skipped tests due to int64 not supported

* torch_load with Tensors (#8037)

* torch_load with Tensors

* remove passthrough_reset + use accept_filename

* Revert "remove passthrough_reset"

* version note

* cleanup

* move copy into the JIT for openpilot compile3 (#7937)

* move copy into the JIT, test fails

* ahh, prune was the issue

* np tensors have the memory from numpy in compile3 [pr] (#8098)

* assign early folding (#8093)

* assign early folding [pr]

* move to to_si

* -

* fix generate_dataset

* diff too big

* no recreation, no diff

* gzip

* new sops from tiny10

* final try

* VALID early folding (#8100)

* fold valid

* :)

* fix test_verify_ast

* keep symbolic working

* nv hcq queue touchups (#8102)

* failing test for const folding breaking indexing [pr] (#8103)

* add avg pool 3d failure test (#8105)

* add test

* try simplify test case

* add TODO comment

* clean up bounds in Tensor.shard (#8107)

* hcq buffer is a class (#8106)

* hcq buffer is a class

* qcom

* no from_mv in qcom

* remove qcombuffer

* useless cast

* mypy

* qcom fix

* _md -> meta

* hcq remove update from args state (#8104)

* hcq remove update from args state

fix amd

ugh

qcom?

qcom ops

ops

qcom fix

qcom texture info

fx

qcom fix

qcom

qcom, sry

minor

works

* remove old code

* unrelated+sint

* qcom

* typing

* rm comments

* set NOOPT in test_avg_pool3d_failure (#8112)

* set NOOPT=0 in test_avg_pool3d_failure

* noopt should still pass

* revert VALID early folding for now (#8114)

This reverts commit 4074f523179331d3a99dd55fa4ef14946563f246.

* init noop changes from delete_lazy [pr] (#8115)

* fix var vmax inside special (#8116)

* fix bitwise_not for signed int (#8117)

-1 is correct because 2**32-1 is not within int32 range, so in some case clang casts the whole thing into uint32

* fix argmax/min on int32 min (#8118)

* test_pickle_realized_tensor actually tests pickle [pr] (#8119)

* test_pickle_realized_tensor actually tests pickle [pr]

* clang

* Remove f16 decompression lib from SD compile.py (#8121)

* Remove f16-to-f32-gpu lib, use tinygrad exported decompression

* No need to create new instance

* hotfix: default llama3 is 1B with download_model

* hotfix: add --size 8B to llama3

* scheduler ops_folding from delete_lazy (#8124)

* scheduler diff from delete_lazy

* test_std_mean

* late fold copy of CONST

* clang const is fine

---------

Co-authored-by: chenyu <chenyu@fastmail.com>
Co-authored-by: George Hotz <geohot@gmail.com>
Co-authored-by: George Hotz <72895+geohot@users.noreply.github.com>
Co-authored-by: qazal <77887910+Qazalin@users.noreply.github.com>
Co-authored-by: Sieds Lykles <93992551+S-Lykles@users.noreply.github.com>
Co-authored-by: Ahmed Harmouche <ahmedharmouche92@gmail.com>
Co-authored-by: ignaciosica <p4ssenger.developer@gmail.com>
Co-authored-by: nimlgen <138685161+nimlgen@users.noreply.github.com>
Co-authored-by: uuuvn <83587632+uuuvn@users.noreply.github.com>
Co-authored-by: leopf <43857362+leopf@users.noreply.github.com>
Co-authored-by: geohotstan <135171913+geohotstan@users.noreply.github.com>
Co-authored-by: Francis Lata <francisalbertlata@gmail.com>
Co-authored-by: JaSpa99 <jakob@craalse.de>
Co-authored-by: mesozoic-egg <133102390+mesozoic-egg@users.noreply.github.com>
Co-authored-by: ttomsa <tomasvsilva8@gmail.com>
